relay/api/resources.py

```diff
@@ -416,24 +409,51 @@ def post(self, args, network_address: str):
             max_fees=max_fees,
             max_hops=max_hops)
 
-        if path:
-            try:
-                gas = self.trustlines.currency_network_proxies[network_address].estimate_gas_for_transfer(
-                    source,
-                    source,
-                    value,
-                    cost,  # max_fee for smart contract
-                    path[1:])  # the smart contract takes the sender of the message as source
-            except ValueError as e:  # should mean out of gas, so path was not right.
-                gas = 0
-                path = []
-                cost = 0
-        else:
-            gas = 0
+        return _estimate_gas_for_transfer(self.trustlines, PaymentPath(cost, path, value), network_address)
+
+
+# CloseTrustline is similar to the above ReduceDebtPath, though it does not
+# take `via` and `value` as parameters. Instead it tries to reduce the debt to
+# zero and uses any contact to do so.
+
+
+class CloseTrustline(Resource):
+
+    def __init__(self, trustlines: TrustlinesRelay) -> None:
+        self.trustlines = trustlines
+
+    args = {
+        'maxHops': fields.Int(required=False, missing=None),
+        'maxFees': fields.Int(required=False, missing=None),
+        'from': custom_fields.Address(required=True),
+        'to': custom_fields.Address(required=True),
+    }
+
+    @use_args(args)
+    @dump_result_with_schema(PaymentPathSchema())
+    def post(self, args, network_address: str):
+        abort_if_unknown_network(self.trustlines, network_address)
+        source = args['from']
+        target_reduce = args['to']
+        max_fees = args['maxFees']
+        max_hops = args['maxHops']
+
+        graph = self.trustlines.currency_network_graphs[network_address]
+        balance = graph.get_balance(source, target_reduce)
+        value = -balance
+
+        if balance == 0:
+            return PaymentPath(fee=0, path=[], value=value, estimated_gas=0)
+
+        if balance > 0:
+            raise RuntimeError("balance is positive. Cannot reduce debt in CloseTrustline resource.")
```

Comment:
return client error code instead.. I think 409 conflict is probably best.

Reply:
either with abort or own exception

Reply:
I think 501 not implemented is more appropriate. Where's the conflict?

Reply:
We can also go with 501. It depends if we think it is our fault that it is not implemented or theirs for using it in the wrong way. 
The conflict would be that the network is not in a state that it should be when calling this endpoint.